### PR TITLE
Support Image flavors

### PR DESCRIPTION
**Pull request by dliappis**
First of all, apologies for the length of this PR.
I tried to keep the commit messages explanatory but the TL;DR is that with this merged we'll be able to build `-oss`, `-basic` and `-platinum` images.

The core element if the `IMAGE_FLAVORS` make variable. For now it defaults `oss platinum` but `basic` will come soon in x-pack.

We should rely on the defaults, but it is possible (and handy) to be able to `IMAGE_FLAVORS="oss platinum" make`. The usual `ELASTIC_VERSION`, `STAGING_BUILD_NUM` can be overridden as usual too.

Examples of produced docker images are:

- `docker.elastic.co/elasticsearch/elasticsearch:6.0.0-beta1-oss`
- `docker.elastic.co/elasticsearch/elasticsearch:6.0.0-beta1-basic`
- `docker.elastic.co/elasticsearch/elasticsearch:6.0.0-beta1-platinum`

In addition to the above the make variable `DEFAULT_IMAGE_FLAVOR` controls which of those will also be tagged with the plain version. Now it has the value `oss` now, which means that `docker.elastic.co/elasticsearch/elasticsearch:6.0.0-beta1` -> `docker.elastic.co/elasticsearch/elasticsearch:6.0.0-beta1-oss`

Other highlights:

- The Makefile targets remain dummy and rely on a simple loop over `$(IMAGE_FLAVORS)`. Note this "make" specific looping, not bash.
  - The original plan of relying on standard Makefile dynamic file dependencies proved too complex. For example extracting stems from things like `tests/docker-compose-%.yml` using [automatic variables](https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html) isn't trivial or very readable. Additionally controlling which image flavors to build through an env var is beneficial.
- Each image gets built from a template generated `Dockerfile-<flavor>` under `build/elasticsearch`.
- `pytest` requires a mandatory parameter `--image-flavor` arg to skip tests not valid for a specific image.
- `./docker-compose.yml` has been replaced by template generated `./docker-compose-<flavor>.yml` files.
- Likewise, the fragment `tests/docker-compose.yml` has been replaced by template generated `tests/docker-compose-<flavor>.yml` files, since the env vars for the fixtures should be image flavor specific (and some can cause start up problems, say, when specifying x-pack parameters for the -oss image.)
- `templates/Dockerfile.j2` has more conditionals to skip x-pack and security specific stuff depending on the image getting built. I opted for more duplication for the sake of being able to see more easily the conditional sections for the various Dockerfile `RUN` commands. The problem is exacerbated by the wish to keep also the resulting `Dockerfile` neat, with `\` separators for long lines, which can become problematic with j2 conditional sections; at the very least there can be multiple LF characters, in the worst case double `\\` characters messing up `RUN` commands.
- Refactoring of `release-manager*:` targets. I added the tests as a dependency, but probably there was a reason they weren't included in the first place so we could discuss this. Now `VERSIONED_IMAGE` is conditionally set to a default, however, I wonder if we could just use `ELASTIC_VERSION` instead for the recursive invocation of `make build-from-local-artifacts tests-only`
- I added some visual cues with `pyfiglet` during tests and release-manager builds, as commands inside make targets merged with `\;` show up as a single line making debugging hard, if the executing command doesn't display the parameters (one such example is `pytest`).

I tested builds, tests + push using BC builds (setting `ELASTIC_VERSION` and `STAGING_BUILD_NUM`), release-manager and release-manager-snapshot builds from 6.0 and 6.x branches (through `ARTIFACTS_DIR`) as well just `ELASTIC_VERSION=6.0.0-beta1 make` successfully.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,61 +1,103 @@
...
 
-test: lint build docker-compose.yml
-	./bin/pytest tests
-	./bin/pytest --single-node tests
+test: lint build docker-compose test-only
+
+# Test specified versions without building
+test-only: lint docker-compose
+	$(foreach FLAVOR, $(IMAGE_FLAVORS), \
+	pyfiglet -w 160 -f puffy "test: $(FLAVOR) image"; \
```
Comment (jarpy) [inline Makefile]: Adorable _and_ useful. 💚

Why it's not called "Pyglet" remains a mystery for the ages.

File: Makefile [hunk 2]
```diff
@@ -1,61 +1,103 @@
...
-	docker rmi -f $(VERSIONED_IMAGE)
+	-$(foreach FLAVOR, $(IMAGE_FLAVORS), \
+	docker rmi -f $(VERSIONED_IMAGE)-$(FLAVOR); \
+	)
+	-docker rmi -f $(VERSIONED_IMAGE)
 
+# Give us an easy way to start the DEFAULT_IMAGE_FLAVOR
 run: run-single
 
-run-single: build docker-compose.yml
```
Comment (jarpy) [inline Makefile]: Another non-phony target bites the dust. Such is progress.

File: templates/Dockerfile.j2 [hunk 3]
```diff
@@ -41,38 +41,80 @@ RUN set -ex && for esdirs in config data logs; do \
...
-{% endif-%}
+{%- endif %}
+
+USER root
+
+{# A fair bit of duplication here, but making line breaks and '\' consistent
+   coupled with jinja2 conditionals is complicated.
+-#}
 
+{% if 'oss' in image_flavor -%}
```
Comment (dliappis) [inline templates/Dockerfile.j2]: Addressed in 21461bf6cb53d6b8659140e2770c75a6fdb34533

File: templates/Dockerfile.j2 [hunk 4]
```diff
@@ -41,38 +41,80 @@ RUN set -ex && for esdirs in config data logs; do \
...
       config/elasticsearch.yml \
       config/log4j2.properties \
       config/x-pack/log4j2.properties \
       bin/es-docker && \
     chmod 0750 bin/es-docker
+{% endif -%}
 
 USER elasticsearch
 
+{% if image_flavor=='platinum' -%}
```
Comment (dliappis) [inline templates/Dockerfile.j2]: Addressed in 21461bf6cb53d6b8659140e2770c75a6fdb34533

File: templates/docker-compose-fragment.yml.j2 [hunk 5]
```diff
@@ -34,8 +39,13 @@ services:
   elasticsearch2:
     hostname: elasticsearch2
     environment:
+      {% if image_flavor == 'platinum' -%}
       - ELASTIC_PASSWORD=pleasechangeme
       - ALLOW_INSECURE_DEFAULT_TLS_CERT=true
+      {% endif -%}
+      {% if 'oss' not in image_flavor -%}
```
Comment (dliappis) [inline templates/docker-compose-fragment.yml.j2]: Addressed in 21461bf6cb53d6b8659140e2770c75a6fdb34533

File: Makefile [hunk 6]
```diff
@@ -1,61 +1,103 @@
...
+endif
+
+# Which image flavor will additionally receive the plain `:version` tag
+ifndef DEFAULT_IMAGE_FLAVOR
+  DEFAULT_IMAGE_FLAVOR := oss
+endif
+
+ifndef VERSIONED_IMAGE
+  VERSIONED_IMAGE := $(ELASTIC_REGISTRY)/elasticsearch/elasticsearch:$(VERSION_TAG)
+endif
```
Comment (dliappis) [inline Makefile]: Addressed in 75e8c5f0755d7a3c38546376b5223671f8d1f976

File: Makefile [hunk 7]
```diff
@@ -1,61 +1,103 @@
...
 all: build test
 
-test: lint build docker-compose.yml
-	./bin/pytest tests
-	./bin/pytest --single-node tests
+test: lint build docker-compose test-only
+
+# Test specified versions without building
+test-only: lint docker-compose
+	$(foreach FLAVOR, $(IMAGE_FLAVORS), \
```
Comment (dliappis) [inline Makefile]: Addressed in 41ee6d438e45b40429d13df20b24a4cfa83dd1fd

File: templates/Dockerfile.j2 (hunk 4 above)
Comment (dliappis) [inline templates/Dockerfile.j2]: and in 97e034012d5a36b824d93943e43afbd99da36ddc and a9bf1a67069333dc5276e94964f62d7ab975da47.

File: templates/docker-compose-fragment.yml.j2 (hunk 5 above)
Comment (dliappis) [inline templates/docker-compose-fragment.yml.j2]: and in 97e034012d5a36b824d93943e43afbd99da36ddc and a9bf1a67069333dc5276e94964f62d7ab975da47.

**Pull request merged**